File: en/codes/python/chapter_stack_and_queue/stack.py
"""
File: stack.py
Created Time: 2022-11-29
Author: Peng Chen (pengchzn@gmail.com)
"""

"""Driver Code"""
if __name__ == "__main__":
    # Initialize stack
    # Python does not have a built-in stack class, but you can use a list as a stack
    stack: list[int] = []

    # Element push
    stack.append(1)
    stack.append(3)
    stack.append(2)
    stack.append(5)
    stack.append(4)
    print("Stack stack =", stack)

    # Access stack top element
    peek: int = stack[-1]
    print("Stack top element peek =", peek)

    # Element pop
    pop: int = stack.pop()
    print("Popped element pop =", pop)
    print("Stack after pop =", stack)

    # Get the length of the stack
    size: int = len(stack)
    print("Stack length size =", size)

    # Determine if it's empty
    is_empty: bool = len(stack) == 0
    print("Is the stack empty =", is_empty)

File: ja/codes/cpp/chapter_stack_and_queue/stack.cpp
/**
 * File: stack.cpp
 * Created Time: 2022-11-28
 * Author: qualifier1024 (2539244001@qq.com)
 */

#include "../utils/common.hpp"

/* Driver Code */
int main() {
    /* スタックを初期化 */
    stack<int> stack;

    /* 要素をプッシュ */
    stack.push(1);
    stack.push(3);
    stack.push(2);
    stack.push(5);
    stack.push(4);
    cout << "スタック stack = ";
    printStack(stack);

    /* スタックトップの要素にアクセス */
    int top = stack.top();
    cout << "トップ要素 top = " << top << endl;

    /* 要素をポップ */
    stack.pop(); // 戻り値なし
    cout << "取り出した要素 pop = " << top << "、取り出し後の stack = ";
    printStack(stack);

    /* スタックの長さを取得 */
    int size = stack.size();
    cout << "スタックの長さ size = " << size << endl;

    /* 空かどうかを判定 */
    bool empty = stack.empty();
    cout << "スタックが空かどうか = " << empty << endl;

    return 0;
}
